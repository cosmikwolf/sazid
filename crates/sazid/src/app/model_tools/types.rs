use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct FunctionProperties {
  #[serde(skip)]
  pub name: String,
  #[serde(skip)]
  pub required: bool,
  #[serde(rename = "type")]
  pub property_type: String,
  pub description: Option<String>,
  // #[serde(skip_serializing_if = "Option::is_none")]
  // pub properties: Option<Box<FunctionProperties>>,
  #[serde(rename = "enum", default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub enum_values: Option<Vec<String>>,
}

// #[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
// pub enum PropertyType {
//   #[serde(rename = "boolean")]
//   Boolean,
//   #[serde(rename = "number")]
//   Number,
//   #[serde(rename = "string")]
//   String,
//   #[serde(rename = "pattern")]
//   Pattern,
//   #[serde(rename = "array")]
//   Array,
//   #[serde(rename = "object")]
//   Object(FunctionProperties),
//   #[serde(rename = "null")]
//   Null,
// }

#[derive(Serialize, Deserialize, Debug)]
pub struct FunctionParameters {
  #[serde(rename = "type")]
  pub param_type: String,
  pub required: Vec<String>,
  pub properties: std::collections::HashMap<String, FunctionProperties>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ToolCall {
  pub name: String,
  pub description: Option<String>,
  pub parameters: Option<FunctionParameters>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Commands {
  pub commands: Vec<ToolCall>,
}
