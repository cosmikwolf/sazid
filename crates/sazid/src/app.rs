use arc_swap::{access::Map, ArcSwap};
use futures_util::Stream;
use helix_core::{diagnostic::Severity, pos_at_coords, syntax, Selection};
use helix_lsp::{
  lsp::{self, notification::Notification},
  util::lsp_range_to_range,
  LspProgressMap,
};
use helix_stdx::path::get_relative_path;
use helix_view::{
  align_view,
  document::DocumentSavedEventResult,
  editor::{ConfigEvent, EditorEvent},
  graphics::Rect,
  theme,
  tree::Layout,
  Align, Editor,
};
use serde_json::json;
use tui::backend::Backend;

// use helix_term::{
//   args::Args,
//   config::Config,
//   keymap::Keymaps,
//   ui::{self, overlay::overlaid},
// };

use crate::compositor::{Component, Compositor, Event};
use crate::job::Jobs;

use log::{debug, error, info, warn};
#[cfg(not(feature = "integration"))]
use std::io::stdout;
use std::{collections::btree_map::Entry, io::stdin, path::Path, sync::Arc};

use anyhow::{Context, Error};

use crossterm::{event::Event as CrosstermEvent, tty::IsTty};
#[cfg(not(windows))]
use {signal_hook::consts::signal, signal_hook_tokio::Signals};
#[cfg(windows)]
type Signals = futures_util::stream::Empty<()>;

#[cfg(not(feature = "integration"))]
use tui::backend::CrosstermBackend;

#[cfg(feature = "integration")]
use tui::backend::TestBackend;

#[cfg(not(feature = "integration"))]
type TerminalBackend = CrosstermBackend<std::io::Stdout>;

#[cfg(feature = "integration")]
type TerminalBackend = TestBackend;

type Terminal = tui::terminal::Terminal<TerminalBackend>;

use crossterm::event::KeyEvent;
use helix_view::graphics::CursorKind;
use serde::{Deserialize, Serialize};
use tokio::sync::mpsc;

pub mod color_math;
pub mod consts;
pub mod database;
pub mod errors;
pub mod functions;
pub mod gpt_interface;
pub mod helpers;
pub mod lsp_interface;
pub mod markdown;
pub mod messages;
pub mod request_validation;
pub mod session_config;
pub mod session_view;
pub mod tools;
pub mod treesitter;
pub mod types;

use crate::{
  action::Action,
  components::{home::Home, session::Session},
  config::Config,
  sazid_tui,
};
use tui;

use self::{
  database::data_manager::{add_session, DataManager},
  errors::SazidError,
  session_config::SessionConfig,
};

#[derive(
  Default, Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize,
)]
pub enum Mode {
  #[default]
  Home,
}

pub struct App {
  compositor: Compositor,
  terminal: Terminal,

  config: Arc<ArcSwap<Config>>,

  #[allow(dead_code)]
  theme_loader: Arc<theme::Loader>,
  #[allow(dead_code)]
  syn_loader: Arc<syntax::Loader>,

  signals: Signals,
  jobs: Jobs,
  lsp_progress: LspProgressMap,
  pub cursor: CursorKind,
  pub tick_rate: f64,
  pub frame_rate: f64,
  pub components: Vec<Box<dyn Component>>,
  pub should_quit: bool,
  pub should_suspend: bool,
  pub mode: Mode,
  pub last_tick_key_events: Vec<KeyEvent>,
}

pub fn add_session_sync(
  db_url: &str,
  config: SessionConfig,
) -> Result<database::types::QueryableSession, SazidError> {
  use diesel::prelude::{
    Connection, ExpressionMethods, RunQueryDsl, SelectableHelper,
  };
  let mut conn =
    diesel_async::async_connection_wrapper::AsyncConnectionWrapper::<
      diesel_async::AsyncPgConnection,
    >::establish(db_url)
    .unwrap();
  use crate::app::database::schema::sessions;
  let config = diesel_json::Json::new(config);
  let session = diesel::insert_into(sessions::table)
    // .values((dsl::model.eq(model.to_string()), dsl::rag.eq(rag)))
    .values(sessions::config.eq(&config))
    .returning(database::types::QueryableSession::as_returning())
    .get_result::<database::types::QueryableSession>(&mut conn)?;
  Ok(session)
}

impl App {
  pub async fn new(
    tick_rate: f64,
    frame_rate: f64,
    data_manager: DataManager,
    config: Config,
    syn_loader_conf: syntax::Configuration,
  ) -> Result<Self, SazidError> {
    let home = Home::new();
    let db_url = data_manager.get_database_url();
    let session: Session =
      add_session(&db_url, config.session_config.clone()).await.unwrap().into();
    let mode = Mode::Home;

    let mut theme_parent_dirs = vec![helix_loader::config_dir()];
    theme_parent_dirs.extend(helix_loader::runtime_dirs().iter().cloned());
    let theme_loader =
      std::sync::Arc::new(theme::Loader::new(&theme_parent_dirs));

    let syn_loader = std::sync::Arc::new(syntax::Loader::new(syn_loader_conf));

    #[cfg(not(feature = "integration"))]
    let backend = CrosstermBackend::new(stdout(), &config.editor);

    #[cfg(feature = "integration")]
    let backend = TestBackend::new(120, 150);

    let terminal = Terminal::new(backend)?;
    let area = terminal.size().expect("couldn't get terminal size");
    let mut compositor = Compositor::new(area);
    let config = Arc::new(ArcSwap::from_pointee(config));

    #[cfg(windows)]
    let signals = futures_util::stream::empty();
    #[cfg(not(windows))]
    let signals = Signals::new([
      signal::SIGTSTP,
      signal::SIGCONT,
      signal::SIGUSR1,
      signal::SIGTERM,
      signal::SIGINT,
    ])
    .context("build signal handler")
    .unwrap();

    Ok(Self {
      tick_rate,
      frame_rate,
      components: vec![
        Box::new(home),
        Box::new(session),
        Box::new(data_manager),
      ],
      should_quit: false,
      should_suspend: false,
      cursor: CursorKind::Block,
      mode,
      last_tick_key_events: Vec::new(),

      compositor,
      terminal,

      config,

      theme_loader,
      syn_loader,

      signals,
      jobs: Jobs::new(),
      lsp_progress: LspProgressMap::new(),
    })
  }

  pub async fn run(&mut self) -> Result<(), SazidError> {
    let (action_tx, mut action_rx) = mpsc::unbounded_channel();

    let mut tui = sazid_tui::Tui::new().unwrap();
    tui.tick_rate(self.tick_rate);
    tui.frame_rate(self.frame_rate);
    tui.mouse(true);
    tui.enter().unwrap();

    for component in self.components.iter_mut() {
      component.register_action_handler(action_tx.clone()).unwrap();
    }

    for component in self.components.iter_mut() {
      component.register_config_handler(self.sazid_config.clone()).unwrap();
    }

    for component in self.components.iter_mut() {
      component.init(tui.size().unwrap()).unwrap();
    }

    loop {
      if let Some(e) = tui.next().await {
        match e {
          sazid_tui::Event::Quit => action_tx.send(Action::Quit).unwrap(),
          sazid_tui::Event::Tick => action_tx.send(Action::Tick).unwrap(),
          sazid_tui::Event::Render => action_tx.send(Action::Render).unwrap(),
          sazid_tui::Event::Resize(x, y) => {
            action_tx.send(Action::Resize(x, y)).unwrap()
          },
          sazid_tui::Event::Key(key) => {
            if let Some(keymap) = self.sazid_config.keybindings.get(&self.mode)
            {
              if let Some(action) = keymap.get(&vec![key]) {
                log::info!("Got action: {action:?}");
                action_tx.send(action.clone()).unwrap();
              } else {
                // If the key was not handled as a single key action,
                // then consider it for multi-key combinations.
                self.last_tick_key_events.push(key);

                // Check for multi-key combinations
                if let Some(action) = keymap.get(&self.last_tick_key_events) {
                  log::info!("Got action: {action:?}");
                  action_tx.send(action.clone()).unwrap();
                }
              }
            };
          },
          _ => {},
        }
        for component in self.components.iter_mut() {
          if let Some(action) =
            component.handle_events(Some(e.clone())).unwrap()
          {
            action_tx.send(action).unwrap();
          }
        }
      }

      while let Ok(action) = action_rx.try_recv() {
        if action != Action::Tick && action != Action::Render {
          //          log::debug!("{action:.unwrap()}");
        }
        let cursor_kind = tui.cursor_kind();
        let cursor_pos =
          if let Ok(cursor) = tui.get_cursor() { Some(cursor) } else { None };

        match action {
          Action::Tick => {
            self.last_tick_key_events.drain(..);
          },
          Action::Quit => self.should_quit = true,
          Action::Suspend => self.should_suspend = true,
          Action::Resume => self.should_suspend = false,
          Action::Resize(w, h) => {
            //trace_dbg!("Action::Resize");
            tui.resize(Rect::new(0, 0, w, h)).unwrap();

            let buf = tui.current_buffer_mut();
            for component in self.components.iter_mut() {
              let r = component.draw(buf);
              if let Err(e) = r {
                action_tx
                  .send(Action::Error(format!("Failed to draw: {:?}", e)))
                  .unwrap();
              }
            }

            tui.draw(cursor_pos, cursor_kind).expect("Failed to draw")
          },
          Action::Render => {
            //trace_dbg!("Action::Render");
            let buf = tui.current_buffer_mut();
            for component in self.components.iter_mut() {
              let r = component.draw(buf);
              if let Err(e) = r {
                action_tx
                  .send(Action::Error(format!("Failed to draw: {:?}", e)))
                  .unwrap();
              }
            }
          },
          _ => {},
        }
        for component in self.components.iter_mut() {
          if let Some(action) = component.update(action.clone()).unwrap() {
            action_tx.send(action).unwrap()
          };
        }
      }
      if self.should_suspend {
        tui.suspend().unwrap();
        action_tx.send(Action::Resume).unwrap();
        tui = sazid_tui::Tui::new().unwrap();
        tui.tick_rate(self.tick_rate);
        tui.frame_rate(self.frame_rate);
        tui.mouse(true);
        tui.enter().unwrap();
      } else if self.should_quit {
        tui.stop().unwrap();
        break;
      }
    }
    tui.exit().unwrap();
    Ok(())
  }
}
