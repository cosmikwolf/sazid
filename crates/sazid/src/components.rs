use core::result::Result;
// This is a sample comment
use helix_view::graphics::Rect;
use tokio::sync::mpsc::UnboundedSender;

use crate::{action::Action, app::errors::SazidError, config::Config};

use tui::buffer::Buffer;

pub mod data_manager;
pub mod session;

pub trait Component {
  #[allow(unused_variables)]
  fn register_action_handler(
    &mut self,
    tx: UnboundedSender<Action>,
  ) -> Result<(), SazidError> {
    Ok(())
  }
  #[allow(unused_variables)]
  fn register_config_handler(
    &mut self,
    config: Config,
  ) -> Result<(), SazidError> {
    Ok(())
  }
  fn init(&mut self, _area: Rect) -> Result<(), SazidError> {
    Ok(())
  }
  #[allow(unused_variables)]
  fn update(&mut self, action: Action) -> Result<Option<Action>, SazidError> {
    Ok(None)
  }
  fn draw(&mut self, b: &mut Buffer) -> Result<(), SazidError>;
}
