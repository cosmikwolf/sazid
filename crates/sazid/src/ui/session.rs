use std::{mem::take, num::NonZeroUsize, path::PathBuf, rc::Rc, sync::Arc};

use crate::commands::{self, OnKeyCallback};
use crate::components::session::Session;
use crate::compositor::{Component, Context, Event, EventResult};
use crate::keymap::{macros::*, KeymapResult, Keymaps};
use helix_core::{
  diagnostic::NumberOrString,
  graphemes::{
    ensure_grapheme_boundary_next_byte, next_grapheme_boundary,
    prev_grapheme_boundary,
  },
  movement::Direction,
  syntax::{self, HighlightEvent},
  text_annotations::TextAnnotations,
  unicode::width::UnicodeWidthStr,
  visual_offset_from_block, Change, Position, Range, Selection, Transaction,
};
use helix_view::{
  document::{Mode, SavePoint, SCRATCH_BUFFER_NAME},
  editor::{CompleteAction, CursorShapeConfig},
  graphics::{Color, CursorKind, Margin, Modifier, Rect, Style},
  input::{KeyEvent, MouseButton, MouseEvent, MouseEventKind},
  keyboard::{KeyCode, KeyModifiers},
  Document, Theme, View,
};
use tui::buffer::Buffer as Surface;
use tui::widgets::{Block, Borders, Paragraph, Widget};

use super::{Completion, CompletionItem, ProgressSpinners};

pub struct SessionView {
  pub keymaps: Keymaps,
  on_next_key: Option<OnKeyCallback>,
  pseudo_pending: Vec<KeyEvent>,
  pub(crate) last_insert: (commands::MappableCommand, Vec<InsertEvent>),
  pub(crate) completion: Option<Completion>,
  spinners: ProgressSpinners,
  /// Tracks if the terminal window is focused by reaction to terminal focus events
  terminal_focused: bool,
}

#[derive(Debug, Clone)]
pub enum InsertEvent {
  Key(KeyEvent),
  CompletionApply { trigger_offset: usize, changes: Vec<Change> },
  TriggerCompletion,
  RequestCompletion,
}

impl Default for SessionView {
  fn default() -> Self {
    Self::new(Keymaps::default())
  }
}

impl SessionView {
  pub fn new(keymaps: Keymaps) -> Self {
    Self {
      keymaps,
      on_next_key: None,
      pseudo_pending: Vec::new(),
      last_insert: (commands::MappableCommand::normal_mode, Vec::new()),
      completion: None,
      spinners: ProgressSpinners::default(),
      terminal_focused: true,
    }
  }

  pub fn spinners_mut(&mut self) -> &mut ProgressSpinners {
    &mut self.spinners
  }

  pub fn render_view(
    &self,
    session: &Session,
    doc: &Document,
    view: &View,
    viewport: Rect,
    surface: &mut Surface,
    is_focused: bool,
  ) {
  }

  fn insert_mode(&mut self, cx: &mut commands::Context, event: KeyEvent) {
    if let Some(keyresult) = self.handle_keymap_event(Mode::Insert, cx, event) {
      match keyresult {
        KeymapResult::NotFound => {
          if let Some(ch) = event.char() {
            commands::insert::insert_char(cx, ch)
          }
        },
        KeymapResult::Cancelled(pending) => {
          for ev in pending {
            match ev.char() {
              Some(ch) => commands::insert::insert_char(cx, ch),
              None => {
                if let KeymapResult::Matched(command) =
                  self.keymaps.get(Mode::Insert, ev)
                {
                  command.execute(cx);
                }
              },
            }
          }
        },
        _ => unreachable!(),
      }
    }
  }

  fn command_mode(
    &mut self,
    mode: Mode,
    cxt: &mut commands::Context,
    event: KeyEvent,
  ) {
    match (event, cxt.session.count) {
      // count handling
      (key!(i @ '0'), Some(_)) | (key!(i @ '1'..='9'), _)
        if !self.keymaps.contains_key(mode, event) =>
      {
        let i = i.to_digit(10).unwrap() as usize;
        cxt.session.count = std::num::NonZeroUsize::new(
          cxt.session.count.map_or(i, |c| c.get() * 10 + i),
        );
      },
      // special handling for repeat operator
      (key!('.'), _) if self.keymaps.pending().is_empty() => {
        for _ in 0..cxt.session.count.map_or(1, NonZeroUsize::into) {
          // first execute whatever put us into insert mode
          self.last_insert.0.execute(cxt);
          let mut last_savepoint = None;
          let mut last_request_savepoint = None;
          // then replay the inputs
          for key in self.last_insert.1.clone() {
            match key {
              InsertEvent::Key(key) => self.insert_mode(cxt, key),
              InsertEvent::CompletionApply { trigger_offset, changes } => {
                let (view, doc) = current!(cxt.session);

                if let Some(last_savepoint) = last_savepoint.as_deref() {
                  doc.restore(view, last_savepoint, true);
                }

                let text = doc.text().slice(..);
                let cursor = doc.selection(view.id).primary().cursor(text);

                let shift_position = |pos: usize| -> usize {
                  (pos + cursor).saturating_sub(trigger_offset)
                };

                let tx = Transaction::change(
                  doc.text(),
                  changes.iter().cloned().map(|(start, end, t)| {
                    (shift_position(start), shift_position(end), t)
                  }),
                );
                doc.apply(&tx, view.id);
              },
              InsertEvent::TriggerCompletion => {
                last_savepoint = take(&mut last_request_savepoint);
              },
              InsertEvent::RequestCompletion => {
                let (view, doc) = current!(cxt.session);
                last_request_savepoint = Some(doc.savepoint(view));
              },
            }
          }
        }
        cxt.session.count = None;
      },
      _ => {
        // set the count
        cxt.count = cxt.session.count;
        // TODO: edge case: 0j -> reset to 1
        // if this fails, count was Some(0)
        // debug_assert!(cxt.count != 0);

        // set the register
        cxt.register = cxt.session.selected_register.take();

        self.handle_keymap_event(mode, cxt, event);
        if self.keymaps.pending().is_empty() {
          cxt.session.count = None
        } else {
          cxt.session.selected_register = cxt.register.take();
        }
      },
    }
  }

  #[allow(clippy::too_many_arguments)]
  pub fn set_completion(
    &mut self,
    session: &mut Session,
    savepoint: Arc<SavePoint>,
    items: Vec<CompletionItem>,
    trigger_offset: usize,
    size: Rect,
  ) -> Option<Rect> {
    let mut completion =
      Completion::new(session, savepoint, items, trigger_offset);

    if completion.is_empty() {
      // skip if we got no completion results
      return None;
    }

    let area = completion.area(size, session);
    session.last_completion = Some(CompleteAction::Triggered);
    self.last_insert.1.push(InsertEvent::TriggerCompletion);

    // TODO : propagate required size on resize to completion too
    completion.required_size((size.width, size.height));
    self.completion = Some(completion);
    Some(area)
  }

  pub fn clear_completion(&mut self, session: &mut Session) {
    self.completion = None;
    if let Some(last_completion) = session.last_completion.take() {
      match last_completion {
        CompleteAction::Triggered => (),
        CompleteAction::Applied { trigger_offset, changes } => self
          .last_insert
          .1
          .push(InsertEvent::CompletionApply { trigger_offset, changes }),
        CompleteAction::Selected { savepoint } => {
          let (view, doc) = current!(session);
          doc.restore(view, &savepoint, false);
        },
      }
    }
  }

  pub fn handle_idle_timeout(
    &mut self,
    cx: &mut commands::Context,
  ) -> EventResult {
    commands::compute_inlay_hints_for_all_views(cx.session, cx.jobs);

    EventResult::Ignored(None)
  }

  fn handle_mouse_event(
    &mut self,
    event: &MouseEvent,
    cxt: &mut commands::Context,
  ) -> EventResult {
    if event.kind != MouseEventKind::Moved {
      cxt.session.reset_idle_timer();
    }
    todo!();
    EventResult::Ignored(None)
  }
}

fn viewport_byte_range(
  text: helix_core::RopeSlice,
  row: usize,
  height: u16,
) -> std::ops::Range<usize> {
  // Calculate viewport byte ranges:
  // Saturating subs to make it inclusive zero indexing.
  let last_line = text.len_lines().saturating_sub(1);
  let last_visible_line =
    (row + height as usize).saturating_sub(1).min(last_line);
  let start = text.line_to_byte(row.min(last_line));
  let end = text.line_to_byte(last_visible_line + 1);

  start..end
}

impl Component for SessionView {
  fn handle_event(
    &mut self,
    event: &Event,
    context: &mut Context,
  ) -> EventResult {
    let mut cx = commands::Context {
      session: context.session,
      count: None,
      register: None,
      callback: Vec::new(),
      on_next_key_callback: None,
      jobs: context.jobs,
    };

    match event {
      Event::Paste(contents) => {},
      _ => {},
    }
  }

  fn render(
    &mut self,
    viewport: Rect,
    surface: &mut Surface,
    cx: &mut Context,
  ) {
    let text_style = cx.session.theme.get("ui.text.info");
    let popup_style = cx.session.theme.get("ui.popup.info");

    // Calculate the area of the terminal to modify. Because we want to
    // render at the bottom right, we use the viewport's width and height
    // which evaluate to the most bottom right coordinate.
    let width = self.width + 2 + 2; // +2 for border, +2 for margin
    let height = self.height + 2; // +2 for border
    let area = viewport.intersection(Rect::new(
      viewport.width.saturating_sub(width),
      viewport.height.saturating_sub(height + 2), // +2 for statusline
      width,
      height,
    ));
    surface.clear_with(area, popup_style);

    let block = Block::default()
      .title(self.title.as_str())
      .borders(Borders::ALL)
      .border_style(popup_style);

    let margin = Margin::horizontal(1);
    let inner = block.inner(area).inner(&margin);
    block.render(area, surface);

    Paragraph::new(self.text.as_str()).style(text_style).render(inner, surface);
  }
}
