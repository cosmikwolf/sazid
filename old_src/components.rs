use color_eyre::eyre::Result;
use crossterm::event::{KeyEvent, MouseEvent};
use ratatui::layout::Rect;
use tokio::sync::mpsc::UnboundedSender;

use crate::{
  action::Action,
  tui::{Event, Frame},
};

pub mod app;
pub mod fps;
pub mod ui;
pub mod session;
pub mod gpt_commands;

//// ANCHOR: component
pub trait Component {
  #[allow(unused_variables)]
  fn init(&mut self, tx: UnboundedSender<Action>) -> Result<()> {
    Ok(())
  }
  fn handle_events(&mut self, event: Option<Event>) -> Option<Action> {
    match event {
      Some(Event::Key(key_event)) => self.handle_key_events(key_event),
      Some(Event::Mouse(mouse_event)) => self.handle_mouse_events(mouse_event),
      _ => None,
    }
  }
  #[allow(unused_variables)]
  fn handle_key_events(&mut self, key: KeyEvent) -> Option<Action> {
    None
  }
  #[allow(unused_variables)]
  fn handle_mouse_events(&mut self, mouse: MouseEvent) -> Option<Action> {
    None
  }
  #[allow(unused_variables)]
  fn update(&mut self, action: Action) -> Result<Option<Action>> {
    Ok(None)
  }
  fn draw(&mut self, f: &mut Frame<'_>, rect: Rect);
}
//// ANCHOR_END: component
