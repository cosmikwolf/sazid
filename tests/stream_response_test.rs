extern crate lazy_static;

#[cfg(test)]
mod tests {
  use async_openai::types::Role;
  use ntest::timeout;
  use sazid::action::Action;
  use sazid::app::types::{ChatMessage, ChatResponse, ChatResponseSingleMessage};
  use sazid::components::session::*;
  use tokio::sync::mpsc;

  #[tokio::test]
  #[timeout(10000)]
  pub async fn test_request_response() {
    let mut enter_processing_action_run = false;
    let mut process_response_action_run = false;
    let (tx, mut rx) = mpsc::unbounded_channel::<Action>();
    let mut session = Session::new();
    session.request_response("Hello World".to_string(), tx.clone());
    'outer: loop {
      while let Some(res) = rx.recv().await {
        match res {
          Action::EnterProcessing => {
            enter_processing_action_run = true;
          },
          Action::ProcessResponse(response) => {
            process_response_action_run = true;
            session.response_handler(tx.clone(), response.clone());
            if let ChatResponse::StreamResponse(message) = response {
              insta::assert_yaml_snapshot!(&message, { ".id" => "[id]", ".created"  => "[created]" });
            } else {
              panic!("Expected StreamResponse");
            };
          },
          Action::ExitProcessing => {
            // break;
            if let Some(ChatMessage::ChatCompletionResponseMessage(ChatResponseSingleMessage::StreamResponse(
              combined,
            ))) = session.data.messages.last()
            {
              assert!(process_response_action_run);
              assert!(enter_processing_action_run);
              insta::assert_yaml_snapshot!(&combined, { ".id" => "[id]", ".created"  => "[created]" });
              insta::assert_yaml_snapshot!(&session.data.messages.last().unwrap(), { ".id" => "[id]", ".created"  => "[created]" });
            } else {
              panic!(
                "Expected last transaction message to be StreamResponse {:#?}",
                session.data.messages.last_mut().unwrap()
              );
            }
            break 'outer;
          },
          Action::Update => {},
          Action::Render => {},
          _ => {
            panic!("Unexpected action {:#?}", res);
          },
        }
      }
    }
  }

  #[test]
  fn test_construct_chat_completion_request_message() {
    let mut session = Session::new();
    if let Ok(create_chat_completion_request_message_result) = session.add_chunked_chat_completion_request_messages(
      "testing testing 1 2 3",
      "sazid testing",
      Role::User,
      session.config.model.clone().as_ref(),
      None,
    ) {
      insta::assert_toml_snapshot!(create_chat_completion_request_message_result);
    } else {
      panic!("construct_chat_completion_request_message failed")
    };
  }
}
