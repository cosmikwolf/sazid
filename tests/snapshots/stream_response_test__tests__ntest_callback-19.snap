---
source: tests/stream_response_test.rs
expression: "&chat_message"
---
ChatCompletionResponseMessage:
  StreamResponse:
    - index: 0
      delta:
        role: ~
        content: " information"
        function_call: ~
      finish_reason: ~

